### DISCUSSION TITLE
How to test component responsiveness using react testing library?

### DISCUSSION THREAD

**Discussion by un4uthorized** [chakra-ui/chakra-ui#5558] [Q&A] [answered]
How to test the responsiveness of components that are using object syntax or even useBreakPoint hook?

Ex: I want to test whether or not a component should appear in the mobile version.

I tried to mock the match media, but it can't find the component styles:

Mock:
```
beforeAll(() => {
    Object.defineProperty(window, "matchMedia", {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches: query === "(min-width: 545px)",
        media: query,
        onchange: null,
        addListener: jest.fn(),
        removeListener: jest.fn(),
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
  });
  ```
  Component:
  ```
  <Box
    justifyContent="space-between"
    alignItems={{ xs: "center", md: "flex-start" }}
    flexWrap="wrap"
    display="flex"
    flexDirection={{ base: "row", xs: "column", md: "row" }}
   gap={{ xs: "30px", md: "0px" }}
> 
```

Result:
![2022-02-10 10-19-42](https://user-images.githubusercontent.com/44379716/153416350-21352f26-a732-465b-b69c-175be348af26.png)

**Comment by arcifius** [comment, marked as answer]
There are some cases where you have behaviours that depends on screen resolution ( like showing more/less elements or even different view types ) for those scenarios it would be better if you mock match media with [mock-match-media](https://www.npmjs.com/package/mock-match-media) it provides an integration with Jest and ChakraUI's `useMediaQuery` will play nice with it.

You can also mock `useBreakpointValue` for each test you run and force it to respond as needed.
```javascript
// mockUseBreakpointValue.test.ts

import * as mediaQueryHooks from '@chakra-ui/media-query'

it("renders less components on mobile", () => {
   jest.spyOn(mediaQueryHooks, 'useBreakpointValue').mockImplementationOnce(() => 2) // where 2 can be anything you need
   
   // test code
})
```

But in this particular case you bought here I really thing you should not worry about the output, using responsive styles in ChakraUI components is a feature ChakraUI provides to us and testing third party libraries isn't a best practice you should assume it always works.

That said, if you are using responsive styles to control component visibility it would be nice to use e2e tests to ensure users are receiving the expected experience ( even assuming Chakra will always work it is still possible to pass wrong styles for breakpoints )

**Reply by un4uthorized** [reply]
Thanks for sharing! It worked for me.

**Reply by Chr1s-create** [reply]
@devneto can you share your working test? I have encounter the same problem on my testing. Thank you so much

**Reply by pbas4** [reply]
Could you provide an example please? I am facing the same issue... Thanks!

**Reply by lsfgrd** [reply]
I'm having the same issue, can't seem to get around it

**Comment by lsfgrd** [comment]
I was having the same issue of not being able to see Chakra UI styles when using the responsive object syntax.

This was solved by wrapping my component under test with `ChakraProvider` in a custom render function.

This is basic config for the library but I forgot about it.
